src/main/java/io/javalin/core/JavalinConfig.java

```diff
@@ -162,25 +160,29 @@ public static void applyUserConfig(Javalin app, JavalinConfig config, Consumer<J
             listener.wsHandlerAdded(x -> anyHandlerAdded.set(true));
         });
 
-        config.inner.plugins.values()
-            .stream()
-            .filter(plugin -> plugin instanceof PluginLifecycleInit)
+        getPluginsExtending(config, PluginLifecycleInit.class)
             .forEach(plugin -> {
-                ((PluginLifecycleInit) plugin).init(app);
+                plugin.init(app);
                 if (anyHandlerAdded.get()) { // check if any "init" added a handler
-                    throw new PluginInitLifecycleViolationException(plugin.getClass());
+                    throw new PluginInitLifecycleViolationException(((Plugin) plugin).getClass());
                 }
             });
 
         config.inner.plugins.values().forEach(plugin -> plugin.apply(app));
 
+        config.inner.server = JettyUtil.getOrDefault(config.inner.server);
+        getPluginsExtending(config, PluginLifecycleApplyServer.class)
+            .forEach(plugin -> plugin.apply(config.inner.server));
+
         if (config.enforceSsl) {
             app.before(SecurityUtil::sslRedirect);
         }
-        if (config.metricsProvider == MetricsProvider.MICROMETER) { // only have one at the moment
-            config.inner.server = JettyUtil.getOrDefault(config.inner.server);
-            JavalinMicrometer.init(config.inner.server);
-        }
     }
 
+    private static <T> Stream<? extends T> getPluginsExtending(JavalinConfig config, Class<T> clazz) {
```

Comment:
This method is a bit weird IMO. Either make it a general util function that accepts a collection and a class, or make it non static so it has access to `config.inner.plugins`.

I think the latter is the better approach, so you could do:

```
pluginsExtending(PluginLifecycleInit.class).forEach {
// and
pluginsExtending(PluginLifecycleApplyServer.class).forEach {
```

Reply:
Totally agree